File: src/main/go/leetcode/lists/hot100/LeetCode98.go
package hot100

func isValidBST(root *TreeNode) bool {
	if root == nil {
		return true
	}
	var isValidBST func(root, max, min *TreeNode) bool
	isValidBST = func(root, max, min *TreeNode) bool {
		if root == nil {
			return true
		}
		if max != nil && root.Val >= max.Val {
			return false
		}
		if min != nil && root.Val <= min.Val {
			return false
		}
		return isValidBST(root.Left, root, min) && isValidBST(root.Right, max, root)
	}
	return isValidBST(root, nil, nil)
}

File: src/main/java/leetcode/lists/hot100/LeetCode98.java
package leetcode.lists.hot100;

import common.*;

/**
 * @author wujingxinit@outlook.com
 * @date 2022/9/6 23:58
 */
public class LeetCode98 {

    /*
     *      5
     *   4     6
     *        3 7
     *
     *
     *    2
     * 1     3
     *
     * */

    public boolean isValidBST(TreeNode root) {
        if (root == null) {
            return true;
        }
        //fix null null
        return isValidBST(root, null, null);
    }

    public boolean isValidBST(TreeNode root, TreeNode min, TreeNode max) {
        if (root == null) {
            return true;
        }
        if (min != null) {
            if (root.val <= min.val) {
                return false;
            }
        }
        if (max != null) {
            if (root.val >= max.val) {
                return false;
            }
        }
        return isValidBST(root.left, min, root) && isValidBST(root.right, root, max);
    }
}